File: src/main/java/learn/spring/RESTfulWebServices/p01_DAOModel/UserDAOService.java
package learn.spring.RESTfulWebServices.p01_DAOModel;

import jakarta.validation.Valid;
import learn.spring.RESTfulWebServices.p02_ExceptionHandling.UserNotFoundException;
import org.springframework.hateoas.EntityModel;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.net.URI;
import java.util.List;

@RestController
public class UserDAOService {
    private final UserDAO service;

    public UserDAOService(UserDAO service) {
        this.service = service;
    }

    @GetMapping("/users")
    public List<User> retrieveAllUsers() {
        return service.findAll();
    }

    @GetMapping("/users/{id}")
    public EntityModel<User> retrieveUser(@PathVariable int id) throws UserNotFoundException {
        User user = service.findById(id);
        if (user == null)
            throw new UserNotFoundException("userId=" + id);

        return EntityModel.of(user);
    }


    @DeleteMapping("/users/{id}")
    public void deleteUser(@PathVariable int id) throws UserNotFoundException {
        service.deleteById(id);
    }

    //    @RequestMapping(path = "/users", method = RequestMethod.POST)
    @PostMapping(path = "/users")
    public ResponseEntity<User> createUser(@Valid @RequestBody User user) {
        service.save(user);
        URI location = ServletUriComponentsBuilder
                .fromCurrentRequest()
                .path("/{id}")
                .buildAndExpand(user.getId())
                .toUri();
        return ResponseEntity.created(location).build();
    }
}

File: src/main/kotlin/learn/spring/kotlinRESTfulWebServices/p01_DAOModel/UserDAOService.kt
package learn.spring.kotlinRESTfulWebServices.p01_DAOModel

import learn.spring.RESTfulWebServices.p02_ExceptionHandling.UserNotFoundException
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.support.ServletUriComponentsBuilder
import java.net.URI

@RestController
class UserDAOService(private val service: UserDAO) {

    @GetMapping("/users")
    fun getAllUsers(): List<User> {
        return service.findAll()
    }

    @GetMapping("/users/{id}")
    fun getUser(@PathVariable id: Int): User {
        val user: User =
            service.findById(id)
        if (user == null)
            throw UserNotFoundException("id=$id")
        return user
    }

    @DeleteMapping("/users/{id}")
    fun deleteUser(@PathVariable id: Int) {
        service.deleteById(id)
    }

    @PostMapping("/users")
    fun createUser(@RequestBody user: User): ResponseEntity<User> {
        val newUser: User = service.save(user)
        val location: URI = ServletUriComponentsBuilder
            .fromCurrentRequest()
            .path("/{id}")
            .buildAndExpand(newUser.id)
            .toUri()
        return ResponseEntity.created(location).build()
    }
}